docs/HowToUseSmartFS.md

```diff
@@ -0,0 +1,139 @@
+```
+
+The code directly above the MTD driver is the SMART MTD layer.
+This interfaces with the MTD (flash driver) layer and handles low-level
+media operations such as logical sector allocation, freeing and management,
+erase block management, low-level formatting, wear leveling, etc.
+
+On top of the SMART MTD layer is the Smart Filesystem Layer.
+The SmartFS code uses the logical sector services of the SMART MTD layer to
+provide file and directory level management, such as creating new files,
+chaining logical sectors together to create files, creating directories and
+ file / directory search and management routines.
+
+## FLASH Driver
+Flash driver must provide low-level read/write/erase capability for the FLASH HW.
+Example driver is at os/arch/arm/src/s5j/s5j_sflash.c
+
+During bootup flash driver must be initialized similar to below example snippet.
+```
+void arm_boot(void)
+{
+	...
+	s5j_board_initialize();
+	...
+}
+
+void s5j_board_initialize(void)
+{
+#ifdef CONFIG_S5J_SFLASH
+        s5j_sflash_init();
+#endif
+}
+```
+### MTD Driver
+MTD driver must be enabled to access the underlying FLASH driver.
+Example driver is at os/fs/driver/mtd/mtd_progmem.c.
+
+Below configuration must enabled for using mtd driver.
+```
+CONFIG_MTD=y
+CONFIG_MTD_PARTITION=y
+CONFIG_MTD_PARTITION_NAMES=y
+CONFIG_MTD_PROGMEM=y
+```
+Example initialization is as shown below,
+refer os/arch/arm/src/sidk_s5jt200/src/s5jt200_tash.c
+
+```
+artik053_configure_partitions() -> progmem_initialize ();
+```
+#### SMART MTD Driver
+Smart MTD layer can be enabled using CONFIG_MTD_SMART config
+```
+CONFIG_MTD_SMART=y
+```
+Refer os/fs/driver/mtd/smart.c
+
+smart device can be initialized using as shown below,
+```
+#if defined(CONFIG_MTD_SMART) && defined(CONFIG_FS_SMARTFS)
+                if (!strncmp(types, "smartfs,", 8)) {
+                        char partref[4];
+                        snprintf(partref, sizeof(partref), "p%d", partno);
+                        smart_initialize(CONFIG_SIDK_S5JT200_FLASH_MINOR,
+                                        mtd_part, partref);
+                }
+#endif
+```
+The above snippet creates smart device node similar to /dev/smartp0x.
+For smart device initialization,
+refer os/arch/arm/src/sidk_s5jt200/src/s5jt200_tash.c
+
+##### SMART FS
+Smart FS layer can be enabled by enabling below configurations.
+```
+CONFIG_FS_SMARTFS=y
+CONFIG_SMARTFS_ERASEDSTATE=0xff
+CONFIG_SMARTFS_MAXNAMLEN=32
+CONFIG_SMARTFS_ALIGNED_ACCESS=y
+```
+Below are the optional features configurations.
+Refer os/fs/smartfs/Kconfig for more details
+```
+CONFIG_SMARTFS_MULTI_ROOT_DIRS
+CONFIG_SMARTFS_BAD_SECTOR
+CONFIG_SMARTFS_DYNAMIC_HEADER
+CONFIG_SMARTFS_JOURNALING
+CONFIG_SMARTFS_SECTOR_RECOVERY
+```
+Smart FS can be mounted on smart mtd device.
+Before mounting smart device formatted to smartfs using mksmartfs tool
+Example snippet for formatting and mounting smart fs
+```
+#ifdef CONFIG_SIDK_S5JT200_AUTOMOUNT_USERFS
```

Comment:
1. Is this needed at fs guide doc?
2. I think you had better mention how to mount fs in TASH also.

Reply:
1. Do you mean only #ifdef or the whole snippet?
2. I will add, how to mount fs in TASH

Reply:
1. Yes, automount functionality.